File: backend/src/repositories/in-memory-session.repository.ts
import { LoopSession } from '../models/session';
import { SessionRepository } from './session.repository';

export class InMemorySessionRepository implements SessionRepository {
  private sessions = new Map<string, LoopSession>();

  async listSessions() {
    return Array.from(this.sessions.values()).sort((a, b) =>
      b.updatedAt.localeCompare(a.updatedAt)
    );
  }

  async getSessionById(sessionId: string) {
    return this.sessions.get(sessionId) ?? null;
  }

  async createSession(session: LoopSession) {
    this.sessions.set(session.id, session);

    return session;
  }

  async reset() {
    this.sessions.clear();
  }
}

export const sessionRepository = new InMemorySessionRepository();

File: backend-dotnet/src/Loopr.Api/Repositories/InMemorySessionRepository.cs
using Loopr.Api.Domain;

namespace Loopr.Api.Repositories;

public sealed class InMemorySessionRepository : ISessionRepository
{
    private readonly Dictionary<string, LoopSession> sessions = new(StringComparer.Ordinal);
    private readonly object syncRoot = new();

    public Task<IReadOnlyList<LoopSession>> ListAsync(CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();

        lock (syncRoot)
        {
            return Task.FromResult<IReadOnlyList<LoopSession>>(
                sessions.Values.OrderByDescending(session => session.UpdatedAt).ToArray()
            );
        }
    }

    public Task<LoopSession?> GetByIdAsync(string sessionId, CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();

        lock (syncRoot)
        {
            sessions.TryGetValue(sessionId, out var session);

            return Task.FromResult(session);
        }
    }

    public Task<LoopSession> CreateAsync(
        LoopSession session,
        CancellationToken cancellationToken
    )
    {
        cancellationToken.ThrowIfCancellationRequested();

        lock (syncRoot)
        {
            sessions[session.Id] = session;
        }

        return Task.FromResult(session);
    }

    public Task ResetAsync(CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();

        lock (syncRoot)
        {
            sessions.Clear();
        }

        return Task.CompletedTask;
    }
}
